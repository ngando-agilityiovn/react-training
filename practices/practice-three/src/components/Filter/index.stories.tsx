import { Meta, StoryObj } from '@storybook/react'
import { BrowserRouter } from 'react-router-dom'
import { Accordion, AccordionItem } from '@chakra-ui/react'

// Constants
import { FILTER } from '@/constants'

// Types
import { Title } from '@/types'

// Components
import Filter from '.'

export default {
  title: 'Components/Filter',
  component: Filter,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Accordion>
          <AccordionItem>
            <Story />
          </AccordionItem>
        </Accordion>
      </BrowserRouter>
    ),
  ],
} as Meta

type Story = StoryObj<typeof Filter>

export const Category: Story = {
  args: {
    title: Title.CATEGORY,
    dataFilter: FILTER.category,
  },
}

export const Brand: Story = {
  args: {
    title: Title.BRAND,
    dataFilter: FILTER.brand,
  },
}

export const Size: Story = {
  args: {
    title: Title.SIZE,
    dataFilter: FILTER.size,
  },
}

export const Price: Story = {
  args: {
    title: Title.PRICE,
  },
}
