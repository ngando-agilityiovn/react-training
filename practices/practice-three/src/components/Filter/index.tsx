import {
  AccordionButton,
  AccordionIcon,
  AccordionPanel,
  Box,
  Checkbox,
  CheckboxGroup,
  Flex,
  Text,
  VStack
} from '@chakra-ui/react'
import SliderCustom from '../SliderCustom'

interface IFilter {
  name: string
  quantity: number
}

interface IFilterGroup {
  title: string
  data?: IFilter[]
  handleFilterCategory?: (e: React.ChangeEvent<HTMLInputElement>) => void
  handleFilterBrand?: (e: React.ChangeEvent<HTMLInputElement>) => void
  handleFilterSize?: (e: React.ChangeEvent<HTMLInputElement>) => void
  handleFilterPrice?: (value: number) => void
}

const Filter = ({
  title,
  data,
  handleFilterCategory,
  handleFilterBrand,
  handleFilterSize,
  handleFilterPrice
}: IFilterGroup) => {
  const handleOnChangeCheckbox = (e: React.ChangeEvent<HTMLInputElement>) => {
    handleFilterCategory?.(e)
    handleFilterBrand?.(e)
    handleFilterSize?.(e)
  }

  return (
    <>
      <h2>
        <AccordionButton>
          <Box
            as="span"
            flex="1"
            textAlign="left"
            fontSize="20px"
            color="dark"
            fontWeight="medium"
          >
            {title}
          </Box>
          <AccordionIcon />
        </AccordionButton>
      </h2>
      <AccordionPanel pb={4}>
        {title === 'Price' ? (
          <>
            <SliderCustom onSliderChange={handleFilterPrice} />
          </>
        ) : (
          <CheckboxGroup colorScheme="blue" defaultValue={[]}>
            <VStack display="block">
              {data?.map(({ name, quantity }) => {
                return (
                  <Flex
                    justifyContent="space-between"
                    fontSize="sm"
                    fontWeight="normal"
                    lineHeight="7"
                  >
                    <Checkbox value={name} onChange={handleOnChangeCheckbox}>
                      {name}
                    </Checkbox>
                    <Text>({quantity})</Text>
                  </Flex>
                )
              })}
            </VStack>
          </CheckboxGroup>
        )}
      </AccordionPanel>
    </>
  )
}

export default Filter
