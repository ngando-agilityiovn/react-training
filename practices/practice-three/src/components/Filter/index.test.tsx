import { render, screen, fireEvent } from '@testing-library/react'
import { Accordion, AccordionItem } from '@chakra-ui/react'
import { BrowserRouter } from 'react-router-dom'

// Components
import Filter from '.'

describe('Filter Component', () => {
  // Mock filter functions
  const mockHandleFilterCategory = jest.fn()
  const mockHandleFilterBrand = jest.fn()
  const mockHandleFilterSize = jest.fn()

  // // Mock data for testing
  const testData = [
    { name: 'Option1', quantity: 5 },
    { name: 'Option2', quantity: 8 }
  ]

  it('Render correcty', () => {
    const container = render(
      <BrowserRouter>
        <Accordion>
          <AccordionItem>
            <Filter
              title="Test Filter"
              data={testData}
              handleFilterCategory={mockHandleFilterCategory}
              handleFilterBrand={mockHandleFilterBrand}
              handleFilterSize={mockHandleFilterSize}
            />
          </AccordionItem>
        </Accordion>
      </BrowserRouter>
    )
    expect(container).toMatchSnapshot()
  })

  it('handleCheckboxChange calls filter functions with the correct argument', () => {
    // Render the component with mock functions and data
    render(
      <Accordion>
        <AccordionItem>
          <Filter
            title="Test Filter"
            data={testData}
            handleFilterCategory={mockHandleFilterCategory}
            handleFilterBrand={mockHandleFilterBrand}
            handleFilterSize={mockHandleFilterSize}
          />
        </AccordionItem>
      </Accordion>
    )

    // Find checkboxes and fire change event
    const checkboxes = screen.getAllByTestId('checkbox')
    checkboxes.forEach((checkbox) => {
      fireEvent.click(checkbox)
    })

    // Expect that each filter function is called with the correct argument
    expect(mockHandleFilterCategory).toHaveBeenCalled()
    expect(mockHandleFilterBrand).toHaveBeenCalled()
    expect(mockHandleFilterSize).toHaveBeenCalled()
  })
})
