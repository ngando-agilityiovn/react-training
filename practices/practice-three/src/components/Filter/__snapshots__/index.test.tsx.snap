// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Component Render correcty 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="chakra-accordion css-0"
      >
        <div
          class="chakra-accordion__item css-1fsnuue"
        >
          <h2>
            <button
              aria-controls="accordion-panel-:r0:"
              aria-expanded="false"
              class="chakra-accordion__button css-ubg1nw"
              data-index="0"
              id="accordion-button-:r0:"
              type="button"
            >
              <span
                class="css-u058mz"
              >
                Test Filter
              </span>
              <svg
                aria-hidden="true"
                class="chakra-icon chakra-accordion__icon css-8o5ky2"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </h2>
          <div
            class="chakra-collapse"
            style="overflow: hidden; display: none; opacity: 0; height: 0px;"
          >
            <div
              aria-labelledby="accordion-button-:r0:"
              class="chakra-accordion__panel css-b0qdn7"
              id="accordion-panel-:r0:"
              role="region"
            >
              <div
                class="chakra-stack css-nczau2"
              >
                <div
                  class="css-kky022"
                >
                  <label
                    class="chakra-checkbox css-1elxtiq"
                    data-testid="checkbox"
                  >
                    <input
                      class="chakra-checkbox__input"
                      style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                      type="checkbox"
                      value="Option1"
                    />
                    <span
                      aria-hidden="true"
                      class="chakra-checkbox__control css-dnty2r"
                    />
                    <span
                      class="chakra-checkbox__label css-1oeb2oe"
                    >
                      Option1
                    </span>
                  </label>
                  <p
                    class="chakra-text css-0"
                  >
                    (
                    5
                    )
                  </p>
                </div>
                <div
                  class="css-kky022"
                >
                  <label
                    class="chakra-checkbox css-1elxtiq"
                    data-testid="checkbox"
                  >
                    <input
                      class="chakra-checkbox__input"
                      style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                      type="checkbox"
                      value="Option2"
                    />
                    <span
                      aria-hidden="true"
                      class="chakra-checkbox__control css-dnty2r"
                    />
                    <span
                      class="chakra-checkbox__label css-1oeb2oe"
                    >
                      Option2
                    </span>
                  </label>
                  <p
                    class="chakra-text css-0"
                  >
                    (
                    8
                    )
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="chakra-accordion css-0"
    >
      <div
        class="chakra-accordion__item css-1fsnuue"
      >
        <h2>
          <button
            aria-controls="accordion-panel-:r0:"
            aria-expanded="false"
            class="chakra-accordion__button css-ubg1nw"
            data-index="0"
            id="accordion-button-:r0:"
            type="button"
          >
            <span
              class="css-u058mz"
            >
              Test Filter
            </span>
            <svg
              aria-hidden="true"
              class="chakra-icon chakra-accordion__icon css-8o5ky2"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </button>
        </h2>
        <div
          class="chakra-collapse"
          style="overflow: hidden; display: none; opacity: 0; height: 0px;"
        >
          <div
            aria-labelledby="accordion-button-:r0:"
            class="chakra-accordion__panel css-b0qdn7"
            id="accordion-panel-:r0:"
            role="region"
          >
            <div
              class="chakra-stack css-nczau2"
            >
              <div
                class="css-kky022"
              >
                <label
                  class="chakra-checkbox css-1elxtiq"
                  data-testid="checkbox"
                >
                  <input
                    class="chakra-checkbox__input"
                    style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                    type="checkbox"
                    value="Option1"
                  />
                  <span
                    aria-hidden="true"
                    class="chakra-checkbox__control css-dnty2r"
                  />
                  <span
                    class="chakra-checkbox__label css-1oeb2oe"
                  >
                    Option1
                  </span>
                </label>
                <p
                  class="chakra-text css-0"
                >
                  (
                  5
                  )
                </p>
              </div>
              <div
                class="css-kky022"
              >
                <label
                  class="chakra-checkbox css-1elxtiq"
                  data-testid="checkbox"
                >
                  <input
                    class="chakra-checkbox__input"
                    style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                    type="checkbox"
                    value="Option2"
                  />
                  <span
                    aria-hidden="true"
                    class="chakra-checkbox__control css-dnty2r"
                  />
                  <span
                    class="chakra-checkbox__label css-1oeb2oe"
                  >
                    Option2
                  </span>
                </label>
                <p
                  class="chakra-text css-0"
                >
                  (
                  8
                  )
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
